Occasional StackOverflow in recursive_pickle (PyPy)

```
=================================== FAILURES ===================================
____________________________ test_recursive_pickle _____________________________

    def test_recursive_pickle():
    
        with AllowPickle():
            f = partial(capture)
            f.__setstate__((f, (), {}, {}))
            try:
                for proto in range(pickle.HIGHEST_PROTOCOL + 1):
                    with pytest.raises(RecursionError):
>                       pickle.dumps(f, proto)

tests/test_partial.py:87: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/Users/runner/hostedtoolcache/PyPy/3.5.3/x64/lib-python/3/pickle.py:1573: in _dumps
    _Pickler(f, protocol, fix_imports=fix_imports).dump(obj)
/Users/runner/hostedtoolcache/PyPy/3.5.3/x64/lib-python/3/pickle.py:408: in dump
    self.save(obj)
/Users/runner/hostedtoolcache/PyPy/3.5.3/x64/lib-python/3/pickle.py:520: in save
    self.save_reduce(obj=obj, *rv)
/Users/runner/hostedtoolcache/PyPy/3.5.3/x64/lib-python/3/pickle.py:603: in save_reduce
    save(args)
/Users/runner/hostedtoolcache/PyPy/3.5.3/x64/lib-python/3/pickle.py:475: in save
    f(self, obj) # Call unbound method with explicit self
/Users/runner/hostedtoolcache/PyPy/3.5.3/x64/lib-python/3/pickle.py:729: in save_tuple
    save(element)
/Users/runner/hostedtoolcache/PyPy/3.5.3/x64/lib-python/3/pickle.py:520: in save
    self.save_reduce(obj=obj, *rv)
/Users/runner/hostedtoolcache/PyPy/3.5.3/x64/lib-python/3/pickle.py:603: in save_reduce
    save(args)
/Users/runner/hostedtoolcache/PyPy/3.5.3/x64/lib-python/3/pickle.py:475: in save
    f(self, obj) # Call unbound method with explicit self
/Users/runner/hostedtoolcache/PyPy/3.5.3/x64/lib-python/3/pickle.py:729: in save_tuple
    save(element)
/Users/runner/hostedtoolcache/PyPy/3.5.3/x64/lib-python/3/pickle.py:520: in save
[...]
/Users/runner/hostedtoolcache/PyPy/3.5.3/x64/lib-python/3/pickle.py:729: in save_tuple
    save(element)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pickle._Pickler object at 0x0000000107794598>
obj = iteration_utilities.partial(<function capture at 0x00000001058bbe20>)
save_persistent_id = True

    def save(self, obj, save_persistent_id=True):
        self.framer.commit_frame()
    
        # Check for persistent id (defined by a subclass)
        pid = self.persistent_id(obj)
        if pid is not None and save_persistent_id:
            self.save_pers(pid)
            return
    
        # Check the memo
        x = self.memo.get(id(obj))
        if x is not None:
            self.write(self.get(x[0]))
            return
    
        # Check the type dispatch table
        t = type(obj)
        f = self.dispatch.get(t)
        if f is not None:
            f(self, obj) # Call unbound method with explicit self
            return
    
        # Check private dispatch table if any, or else copyreg.dispatch_table
        reduce = getattr(self, 'dispatch_table', dispatch_table).get(t)
        if reduce is not None:
            rv = reduce(obj)
        else:
            # Check for a class with a custom metaclass; treat as regular class
            try:
                issc = issubclass(t, type)
            except TypeError: # t is not a class (old Boost; see SF #502085)
                issc = False
            if issc:
                self.save_global(obj)
                return
    
            # Check for a __reduce_ex__ method, fall back to __reduce__
>           reduce = getattr(obj, "__reduce_ex__", None)
E           SystemError: <StackOverflow object at 0x103ccbf60>

/Users/runner/hostedtoolcache/PyPy/3.5.3/x64/lib-python/3/pickle.py:493: SystemError
```

--- a/tests/test_partial.py
+++ b/tests/test_partial.py
@@ -15,6 +15,7 @@ import pytest
 import iteration_utilities
 from iteration_utilities._compat import (
     RecursionError, AttributeUnwriteableException)
+from iteration_utilities._utils import IS_PYPY
 
 # Test helper
 import helper_funcs as _hf
@@ -43,9 +44,16 @@ def signature(part):
 
 class AllowPickle:
     def __enter__(self):
+        if IS_PYPY:
+            # PyPy sometimes runs into a StackOverflow with the default
+            # recursion limit. So use something smaller for this test.
+            self._oldlimit = sys.getrecursionlimit()
+            sys.setrecursionlimit(100)
         return self
 
     def __exit__(self, typ, value, tb):
+        if IS_PYPY:
+            sys.setrecursionlimit(self._oldlimit)
         return False
 
 

